### PR TITLE
API-5571 audience config

### PR DESCRIPTION
**Pull request by crolarlibertyva**
- Adds the audience setting in the config for each app category

Config PRs
* https://github.com/department-of-veterans-affairs/lighthouse-oauth-proxy-configs/pull/24
* https://github.com/department-of-veterans-affairs/devops/pull/8765

### CODE REVIEW

**Review by beaugrantham-va — CHANGES_REQUESTED** [review]
Needs corresponding devops config PR.

**Review by beaugrantham-va — CHANGES_REQUESTED** [review]
Linting fails.

**Review by lukemarkwordtlibertyits — COMMENTED** [review]
Looks good, my only comment is that the AuthorizationHandler grabs the Audience from the authorizationServer. This will likely cause issues with the new Issuers. It may be worth using the audience from the config in this check. Also this will be less costly!

File: oauth-proxy/cli.js [hunk 1]
```diff
@@ -157,6 +157,11 @@ function processArgs() {
             required: false,
             default: false,
           },
+          audience: {
+            description: "The audience for the app catetory.",
```
Comment (beaugrantham-va) [inline oauth-proxy/cli.js:161]: s/catetory/category

File: oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/getPatientInfoStrategies/getPatientInfoFromValidateEndpointStrategy.js [hunk 2]
```diff
@@ -1,15 +1,16 @@
 const { rethrowIfRuntimeError } = require("../../../utils");
 class GetPatientInfoFromValidateEndpointStrategy {
-  constructor(validateToken, logger) {
+  constructor(validateToken, logger, audience) {
     this.logger = logger;
     this.validateToken = validateToken;
+    this.audience = audience;
   }
-  async createPatientInfo(tokens, decoded) {
```
Comment (beaugrantham-va) [inline oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/getPatientInfoStrategies/getPatientInfoFromValidateEndpointStrategy.js:7]: Anywhere that calls this function needs to be updated to match the new signature.

Comment (beaugrantham-va) [inline oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/getPatientInfoStrategies/getPatientInfoFromValidateEndpointStrategy.js:7]: A side effect of changing this signature is that you'll need to change the signature of the other classes that implement the same "interface" - getPatientInfoFromLaunchStrategy.js

**Review by beaugrantham-va — CHANGES_REQUESTED** [review]
Can you pull the latest master?

File: oauth-proxy/tests/TokenHandlerTests/tokenHandler.test.js [hunk 3]
```diff
@@ -39,7 +40,7 @@ beforeEach(() => {
   });
   validateToken = buildValidateToken({});
   staticTokens = new Map();
-  next = jest.fn();
+  (audience = "https://sandbox-api.va.gov/services/fhir"), (next = jest.fn());
```
[the remark this answers is not in the document]
Comment (bdgrider-va) [reply oauth-proxy/tests/TokenHandlerTests/tokenHandler.test.js:43]: I think the linter might have done this? Changing it back.

File: oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/getPatientInfoStrategies/getPatientInfoFromLaunchStrategy.js [hunk 4]
```diff
@@ -3,7 +3,7 @@ class GetPatientInfoFromLaunchStrategy {
     this.req = req;
   }
   // eslint-disable-next-line no-unused-vars
```
Comment (beaugrantham-va) [inline oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/getPatientInfoStrategies/getPatientInfoFromLaunchStrategy.js:5]: This eslint directive can be removed now.

**Review by beaugrantham-va — CHANGES_REQUESTED** [review]
Still need to pull in the latest master

File: oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/getPatientInfoStrategies/getPatientInfoFromLaunchStrategy.js (hunk 4 above)
Comment (bdgrider-va) [reply oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/getPatientInfoStrategies/getPatientInfoFromLaunchStrategy.js:5]: It's still needed for the tokens parameter.

Comment (beaugrantham-va) [reply oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/getPatientInfoStrategies/getPatientInfoFromLaunchStrategy.js:5]: You're totally right. My bad.

**Review by lukemarkwordtlibertyits — CHANGES_REQUESTED** [review]
This is kindof a tough one to spot, but I am not sure that the config audience ever gets passed through. See this log on the Authorization Handler.

image

If the audience was passed then there would also be an "expected" field in that message.

I think the same is true in the refresh flow.

**Comment by bdgrider-va** [comment]
> This is kindof a tough one to spot, but I am not sure that the config audience ever gets passed through. See this log on the Authorization Handler.
> 
> image
> 
> If the audience was passed then there would also be an "expected" field in that message.
> 
> I think the same is true in the refresh flow.

Good catch, just fixed this.

image

File: oauth-proxy/oauthHandlers/authorizeHandler.js [hunk 5]
```diff
@@ -125,7 +120,7 @@ const authorizeHandler = async (
  *
  * @returns {Promise<{valid: boolean, error?: string, error_description?: string}>}
  */
-const checkParameters = async (state, aud, issuer, logger, oktaClient) => {
+const checkParameters = async (state, aud, audience, logger) => {
```
Comment (beaugrantham-va) [inline oauth-proxy/oauthHandlers/authorizeHandler.js:123]: These param names (`aud`, `audience`) need to be more descriptive and needs `@param` documentation.

**Review by beaugrantham-va — CHANGES_REQUESTED** [review]

File: oauth-proxy/oauthHandlers/authorizeHandler.js [hunk 6]
```diff
@@ -28,6 +28,7 @@ const authorizeHandler = async (
   oktaClient,
   slugHelper,
   app_category,
+  audience,
```
[the remark this answers is not in the document]
Comment (beaugrantham-va) [reply oauth-proxy/oauthHandlers/authorizeHandler.js:31]: If not, you'll need to update the jsdoc.

File: oauth-proxy/oauthHandlers/authorizeHandler.js [hunk 7]
```diff
@@ -136,26 +131,11 @@ const checkParameters = async (state, aud, issuer, logger, oktaClient) => {
...
-      .then((res) => {
-        serverAudiences = res.audiences;
-      })
-      .catch(() => {
-        logger.error("Unable to get the authorization server.");
-        throw { status: 500 };
-      });
-
-    if (!serverAudiences.includes(aud)) {
+    if (!(aud === audience)) {
```
Comment (beaugrantham-va) [inline oauth-proxy/oauthHandlers/authorizeHandler.js:134]: Is there a difference between `!(aud === audience)` and `aud !== audience`? Not saying I know... genuinely curious.

Comment (bdgrider-va) [reply oauth-proxy/oauthHandlers/authorizeHandler.js:134]: It appears that they are both "identity" comparisons, where both type and value are evaluated. So
"2"!==2 and !("2"===2) both evaluate to true

image

File: oauth-proxy/oauthHandlers/authorizeHandler.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (bdgrider-va) [reply oauth-proxy/oauthHandlers/authorizeHandler.js]: It does appear that is the case, removing the audience parameter now.

File: oauth-proxy/oauthHandlers/authorizeHandler.js (hunk 7 above)
Comment (lukemarkwordtlibertyits) [reply oauth-proxy/oauthHandlers/authorizeHandler.js]: If all things are equal than I think `!==` is preferable. To me its easier to read.

File: oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/getPatientInfoStrategies/getPatientInfoFromValidateEndpointStrategy.js [hunk 8]
```diff
@@ -1,15 +1,16 @@
 const { rethrowIfRuntimeError } = require("../../../utils");
 class GetPatientInfoFromValidateEndpointStrategy {
-  constructor(validateToken, logger) {
+  constructor(validateToken, logger, audience) {
     this.logger = logger;
     this.validateToken = validateToken;
+    this.audience = audience;
```
Comment (lukemarkwordtlibertyits) [inline oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/getPatientInfoStrategies/getPatientInfoFromValidateEndpointStrategy.js:6]: I added a debugging line in this class to check that the audience parameter was being passed in. It is `undefined` my setup is pointing at the v1 validation endpoint so I am not sure why it is passing, still if this is a field of this class, it should be populated.

May be worth looking into whether this field is actually necessary any more?

**Review by lukemarkwordtlibertyits — CHANGES_REQUESTED** [review]

File: oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/getPatientInfoStrategies/getPatientInfoFromValidateEndpointStrategy.js (hunk 8 above)
Comment (bdgrider-va) [reply oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/getPatientInfoStrategies/getPatientInfoFromValidateEndpointStrategy.js:6]: Just fixed this. I had missed a second instantiation of the GetPatientInfoFromValidateEndpointStrategy class in the tokenHandlerClientBuidler.

**Review by lukemarkwordtlibertyits — DISMISSED** [review]
Looks good!

**Review by beaugrantham-va — APPROVED** [review]

**Pull request merged**